### ISSUE TITLE
Discuss syntax for pair-shaped metadata

### ISSUE THREAD

**Issue by tstehr** [tstehr/RecipeMD#47]
Issue spun out of the discussion in PR #46.

There's a wish from some users to add some sort of structured metadata to recipes. Examples of metadata include:

- source
- times (baking, cooking, cooling, resting, active, total, …) (see also #27)
- taste
- vegetarian: yes/no
- nutrition information

Some of these can be incorporated into tags, e.g. by adding the taste as a tag or tagging a recipe as vegeratian. But that doesn't really work for [pair-shaped metadata](https://github.com/tstehr/RecipeMD/pull/46#issuecomment-1324872757) like times or nutrition information. For these the only solution currently is to put them into the recipe description, instructions or tags in an ad-hoc format. This has the drawback that different recipe author are going to choose different formats which is not bad per se but it would be nice to have a consistent format to enable tooling support.

## State of the art

To figure out where recipe authors put data like this, I did an informal "survey" by looking at some cookbooks I own and recipe websites I tend to use:

- Recipe metadata usually occurs in these places (sorted by roughly by number of appearances in my non-representative sample)
  1. before the ingredient list
  2. after the instructions at the bottom of the page
  3. after the ingredient list
  4. after the recipe title before the description
- It's also common to split metadata by type, for example [seriouseats.com](https://seriouseats.com) puts cooking times with the yields before the ingredients, while nutrition information is located below the instructions.

## Requirements for RecipeMD

From the discussion in #46 I extracted the following set of requirements that I'd like to satisfy

The overarching concern of adding a feature like this to RecipeMD is that it should feel natural and fit the way people might author their recipes in Markdown if not guided by the specification or tooling requirements.

- The source text should follow the [markdown design philosophy]:

  > Markdown-formatted document should be publishable as-is, as plain text, without looking like it’s been marked up with tags or formatting instructions.

  This includes using a minimal amount of additional required markup and retaining a natural reading order.

- The output of running the RecipeMD file to a commonmark renderer should look reasonable (e.g. converting to HTML or to PDF for printing via [pandoc]).

- The metadata format should allow tooling to do stuff with the metadata in a way that makes it more useful than ad-hoc metadata in the description.

- If the metadata embedded in RecipeMD uses a pre-existing format, it should fit with RecipeMD in spirit and be easy to understand and author.

[markdown design philosophy]: https://daringfireball.net/projects/markdown/syntax#philosophy
[pandoc]: https://recipemd.org/recommended_tools.html#pandoc

## Approaches

So far two approaches have been suggested. We're open for arguments for/against both of these and also for additional suggestions.

### Extended Tags

In #46 I considered using tagstags as a more generic meta data field, by allowing colon-separated key-value-pairs there:

```markdown
*vegan, summer, calories/serving: 39, protein/serving: 1.3 g, sugar/serving: 5.9 g,
fat/serving: 0.4 g, active time: 45 min, total time: 2h*
```

### Metadata as fenced code blocks

In discussion with @AberDerBart an alternative approach came up. We could allow adding a fenced code block (aka ```-block) where tags and yields are currently located and parse its contents (as YAML or a something like that):

````markdown
*vegan, summer*

```
calories/serving: 39
protein/serving: 1.3 g
sugar/serving: 5.9 g
fat/serving: 0.4 g
active time: 45 min
total time: 2h
```
````

**Comment by rprieto** [comment]
I agree this would be a great feature. The ability to capture and view prep/wait time is very useful for choosing recipes. Other tools like Tandoor even display it prominently at the top of each recipe / in the thumbnail view.

Is something that's on the short term roadmap for RecipeMD?

Both the YAML and bullet lists suggestions look good to me. Bullet lists because they feel natural, and YAML because it reinforces that these are special keywords. A middle ground could be fenced keywords in a list:

```md
- `tags`: summer, fresh
- `yield`: 4 servings, 400g
- `prep`: 10min
- `wait`: 1h
```

(tags and yield can be considered to be metadata as well, in a backwards compatible way)
